core/backend/src/CodeService.ts

```diff
@@ -0,0 +1,521 @@
+   */
+  reserveNextAvailableCode: (arg: CodeService.ReserveNextArgs) => Promise<void>;
+
+  /**
+   * Attempt to reserve an array of the next available codes for a code sequence and scope.
+   * The length of the array determines the number of codes requested. The values for the new codes are returned
+   * in the array, so they can be associated with the supplied GUIDs.
+   * @returns number of codes actually reserved.
+   * @note This will automatically attempt to obtain, perform the operation, and then release the write lock.
+   */
+  reserveNextAvailableCodes: (arg: CodeService.ReserveNextArrayArgs) => Promise<number>;
+
+  /**
+   * Update the properties of a single code.
+   * @note This will automatically attempt to obtain, perform the operation, and then release the write lock.
+   */
+  updateCode: (props: CodeService.UpdatedCode) => Promise<void>;
+
+  /**
+   * Update the properties of an array codes.
+   * @note This will automatically attempt to obtain, perform the operation, and then release the write lock.
+   * @note If you have a set of codes to update, it is considerably more efficient to do them as an array rather than one at a time.
+   * @returns number of codes actually updated.
+   */
+  updateCodes: (arg: CodeService.UpdateCodesArgs) => Promise<number>;
+
+  /** Delete an array of codes by their guids.
+   * @note This will automatically attempt to obtain, perform the operation, and then release the write lock.
+   */
+  deleteCodes: (guid: CodeService.CodeGuid[]) => Promise<void>;
+}
+
+/** @alpha */
+export namespace CodeService {
+  /** @internal */
+  const codeSequences = new Map<string, CodeSequence>();
+
+  /** @internal */
+  export let createForBriefcase: undefined | ((db: BriefcaseDb) => CodeService);
+
+  /** Register an instance of a`CodeSequence` so it can be looked up by name. */
+  export function registerSequence(seq: CodeSequence) { codeSequences.set(seq.sequenceName, seq); }
+
+  /** Get a previously registered `CodeSequence` by its name.
+   * @throws if no sequence by that name was registered.
+  */
+  export function getSequence(name: string): CodeSequence {
+    const seq = codeSequences.get(name);
+    if (!seq)
+      throw new Error("SequenceNotFound", -1, `code sequence ${name} not found`);
+    return seq;
+  }
+
+  /**
+   * Turn a `CodePops` for the briefcase of this CodeService into a `ScopeAndSpec` object for use with a CodeService.
+   * This is necessary because the `spec` member of `CodeProps` refers to the id of a code spec in the iModel, and
+   * the `scope` member refers to the element Id of the scope element in the iModel. This helper function
+   * converts the spec Id to the spec name and looks up the `FederationGuid` of the scope element.
+   */
+  export function makeScopeAndSpec(briefcase: BriefcaseDb, props: CodeProps): CodeService.ScopeAndSpec {
+    const scope = briefcase.elements.getElementProps(props.scope).federationGuid;
```

Comment:
Rather wasteful to query all element properties just to get the guid?

Reply:
is there a better way?

Reply:
`SELECT FederationGuid FROM ...`

Reply:
i added `onlyBaseProperties` to `ElementLoadOptions`